### ISSUE TITLE
Need review of how this is configured

### ISSUE THREAD

**Issue by jbfaden** [hapi-server/server-java#14]
This is my code and every time I go to play with it I'm surprised with how opaque the initial configuration is.  I have a server set up which is serving from /home/jbf/tmp/servers/fun, but for the life of me I can't figure out how it knows to use that location.  I look for hapi_home and HAPI_HOME variables, and can't find either.  The documentation says hapi_home should be used, and internal documentation says it's HAPI_HOME.  

What's more, servers should be reluctant to expose server-side information.  I must have decided to be loose with this, because some is exposed.  Either all should be exposed, or none should be exposed. (This might be because I'm connecting from localhost, but even this is risky.  With proxies this is easy to forget.)

This needs to be reviewed and more clearly documented.

**Comment by jbfaden** [comment]
I've made it so that the configuration happens in just one place, Initialize.java.  The code looks like this, where if it is not set in the servlet context, then we look in environment variables and properties.

```java
    public static String getHapiHome(ServletContext servletContext) {
        String HAPI_HOME= servletContext.getInitParameter("hapi_home");
        if ( HAPI_HOME.equals(Initialize.DEFAULT_HAPI_HOME) ) {
            String x= System.getenv("HAPI_HOME");
            if ( x==null ) {
                x= System.getProperty("HAPI_HOME");
            }
            if ( x!=null ) {
                logger.info("setting HAPI_HOME with environment variable.");
                HAPI_HOME= x;
            }
        }
        return HAPI_HOME;
    }
```

**Comment by rweigel** [comment]
I use *HAPISERVERPATH* and *HAPISERVERHOME*for env variables.
https://github.com/hapi-server/server-nodejs#41-confconfigjson

Should I switch to these? Are there any others?

**Comment by jbfaden** [comment]
I'm not sure there's a lot of benefit to making them the same.  I expect the configuration files to be a certain format, and it might be confusing if we use the same names and then the configuration files are different.

**Comment by rweigel** [comment]
Hmm. I thought we agreed there would be a great benefit to standardizing a core set of configuration features. I can see a scenario when on server codebase is no longer updated and to use a new one that has the same capabilities, one simply changes the installed HAPI server binary. This will also make the comparison of speed, memory usage, etc. very easy for users trying to decide what implementation to go with.

**Comment by jbfaden** [comment]
There are things that each server is going to be good at, not because of how they are written but because of the language and platform they run on top of.  The Java one has features that would be very difficult to implement in the Python one, and vice-versa.

**Comment by rweigel** [comment]
If they are billed as being general purpose, they'll have common capabilities, no?

**Comment by jbfaden** [comment]
I agree we should make them as consistent as possible, but for example I wouldn't expect the Python server to be able to call into native Java code to get data, like the Java server can.  Likewise I can imagine there are things the Python one could do which would be difficult to implement in Java.

**Comment by rweigel** [comment]
Of course.

**Comment by jbfaden** [comment]
I'll go through the configurations and see where we can make the configurations more consistent.  The server was developed in a rush before the meeting last year, and matching the existing configuration was not top of mind, but there will be things I need to describe which don't make sense in other contexts.

**Comment by rweigel** [comment]
Great!

**Comment by jbfaden** [comment]
There are a number of differences, some of which you might want to pick up.  For example, it looks like you can have one command to calculate the catalog, and I allow for a mix of different commands and static entries.  I assume that with your code if a command calculates the catalog, then it will also calculate the info responses.  Is that right?  The way we have data command infer subsetting is the same.  I'll make sure we are using the same macros for start, stop, parameters, but I'm pretty sure we are.

**Comment by rweigel** [comment]
> For example, it looks like you can have one command to calculate the catalog

The catalog can be created inline (static) or by using a command line command or a URL that returns the equivalent of the inline content.

The info for each dataset can be the same.

Probably the documentation is not up-to-date.

You can also set the frequency at which everything is refreshed when the server is running.

**Comment by jbfaden** [comment]
How does the command know if it should be the info or the catalog response?  Also, for the updates, why not simply check file timestamps and always run things?  I guess that could be a default behavior, but if things are sufficiently slow then you could have it cache as long as the configuration hasn't been changed and with a freshness timeout.

**Comment by jbfaden** [comment]
Also I'm really bothered again by the fact that we don't have constraints on ids.  Simon has things with spaces in them, so you might make sure these are handled properly, and describe how the commands will receive them.  (Maybe you simply quote the id, but what if it contains a quote--is that handled too, and so on.)

**Comment by rweigel** [comment]
A full configuration is

https://github.com/hapi-server/server-nodejs/blob/master/metadata/TestData2.0.json

At the top level, there must be a key of one of `catalog_inline`, `catalog_uri`, or `catalog_command`. The URL or command must return an array with objects with keys of `id`, `title`, and `info_*`, where `*` is one of `inline`, `url`, or `command`. The value for `info_*` is an object that is identical to what is returned by `hapi/info?dataset=id`.

So, schematically,

```
"catalog_{inline,uri,command}" : (array of objects or string). Array has form of
  [
    {
      "id": "dataset1",
      "title": "Dataset w/valid HAPI 2.0",
      "info_{inline,url,command}": Object or string.
    },...
  ]
```

Sometimes one does not know when anything is updated, which is the reason for the optional update frequency. In principle, I could use HEAD requests to determine if there has been a change before making a request (but how often? I would not do it every time. Some servers are very slow in responding to metadata requests). One could also check file timestamps for URIs to local files. Anyway, this is something for discussion.

I ran into an issue with the spaces yesterday on a web client. (I've had similar problems with the clients with one of your demo servers).

**Comment by jbfaden** [comment]
Yes, at least spaces are discouraged, but it'll always be difficult for people to remember that they are allowed.

I think for my use cases it was easy enough to always make web requests, but I can imagine cases where that wouldn't work well.  Also HEAD requests require that the upstream service implement them, which is probably also unlikely.  So I think your freshness parameter is probably a good solution, and it's good to have it be explicit.  I'll review my code to see how this worked again.  I know I keep copies of the responses on disk, somewhat for transparency, but also because Nand got into trouble keeping everything in memory, and I wanted a way things could be unloaded when they are not used.
